python/lsst/ap/association/assoc_db_sqlite.py

```diff
@@ -0,0 +1,509 @@
+            DM-11921
+        """
+        for updated_collection_index in updated_dia_collection_indices:
+            dia_object = dia_collection.dia_objects[updated_collection_index]
+            dia_object.dia_object_record.set(
+                'indexer_id',
+                self.indexer.index_points([dia_object.ra], [dia_object.dec])[0])
+            self._store_source_record(
+                dia_object.dia_object_record,
+                self._dia_object_converter)
+            self._store_source_record(
+                dia_object.dia_source_catalog[-1],
+                self._dia_source_converter)
+            self._store_dia_object_source_pair(
+                dia_object.id,
+                dia_object.dia_source_catalog[-1].getId())
+        self._commit()
+
+    def get_dia_objects(self, indexer_indices):
+        """ Retrive the DIAObjects from the database whose indexer indices
+        are with the specified list of indices.
+
+        Retrives a list of DIAObjects that are covered by the pixels with
+        indices, indexer_indices. Use this to retrive the complete DIAOjbect
+        including the associated sources.
+
+        Parameters
+        ----------
+        indexer_indices : array like of ints
+            Pixelized indexer indices from which to load.
+
+        Returns
+        -------
+        list of lsst.ap.association.DIAObjects
+        """
+        output_dia_objects = []
+
+        for row in self._query_dia_objects(indexer_indices):
+            dia_object_record = \
+                self._dia_object_converter.source_record_from_db_row(row)
+            dia_sources = self.get_dia_sources(
+                dia_object_record.getId())
+            output_dia_objects.append(
+                DIAObject(dia_sources, dia_object_record))
+
+        return output_dia_objects
+
+    def _query_dia_objects(self, indexer_indices):
+        """ Query the database for the stored DIAObjects given a set of
+        indices in the indexer.
+
+        Parameters
+        ----------
+        indexer_indices : list of ints
+            Spatial indices in the indexer specifying the area of on the sky
+            to load DIAObjects for.
+
+        Return
+        ------
+        list of tuples
+            Query result containing the values representing DIAObjects
+        """
+        self._db_cursor.execute(
+            "CREATE TEMPORARY TABLE tmp_indexer_indices "
+            "(indexer_id INTEGER PRIMARY KEY)")
+        for indexer_index in indexer_indices:
+            self._db_cursor.execute(
+                "INSERT OR REPLACE INTO tmp_indexer_indices VALUES (?)",
+                (indexer_index,))
+
+        self._db_cursor.execute(
+            "SELECT o.* FROM dia_objects AS o "
+            "INNER JOIN tmp_indexer_indices AS i ON o.indexer_id = i.indexer_id")
+
+        output_rows = self._db_cursor.fetchall()
+
+        self._db_cursor.execute(
+            "DROP TABLE tmp_indexer_indices")
```

Comment:
What would the query management you're suggesting look like? I'm wondering if what you are thinking of isn't already taken care of by the sqlite interface.

Reply:
Ideally, everything from creating the temp table to dropping it would be a single transaction to guarantee that it's handled as a single unit. Many DB systems require that table management commands be their own transactions, so it would be more realistic to have the insert and select statements live in their own transaction.

Most databases default to having each query be a transaction, which in addition to causing atomicity/concurrency problems adds a lot of overhead.

Reply:
Hopefully this addresses what you're worried about. I've talked to @ctslater and worked through the code a bit. I've changed the for loop to executemany which should take care of some of the ugliness. Colin and I also discussed a bit about what is happening with the transactions and it seems that the python sqlite3 package doesn't automatically close transactions on each execute so this function, and the whole class, shouldn't close a transaction until commit is called. Also we have confirmed that the TEMPORARY TABLE shouldn't stick around even the processes is closed prematurely. Below is a code snippet you were looking at before. Does this address your problems?

        self._db_cursor.execute(
            "CREATE TEMPORARY TABLE tmp_indexer_indices "
            "(indexer_id INTEGER PRIMARY KEY)")

        self._db_cursor.executemany(
            "INSERT OR REPLACE INTO tmp_indexer_indices VALUES (?)",
            [(indexer_index,) for indexer_index in indexer_indices])

        self._db_cursor.execute(
            "SELECT o.* FROM dia_objects AS o "
            "INNER JOIN tmp_indexer_indices AS i "
            "ON o.indexer_id = i.indexer_id")

        output_rows = self._db_cursor.fetchall()

        self._db_cursor.execute("DROP TABLE tmp_indexer_indices")
        self._commit()

        return output_rows

Reply:
I assume `executemany` takes care of the performance side of things, but I'm not clear on what happens if something goes wrong -- what does/doesn't get rolled back, @ctslater?

As for the temporary table, I would expect it to last as long as your DB connection. If `_query_dia_objects` ever gets called more than once by the same object, you could run into trouble.

Reply:
Because nothing is committed until the end of the function, sqlite will rollback everything up until the commit before `_query_dia_objects` was ever called. That's assuming that some failure occurs inside of sqlite and python execution continues as normal; I think a much more likely scenario is that python somehow crashes during the function, in which case the temporary table automatically disappears and transactions don't even come into play. So I think this is belt-and-suspenders safe when it comes to not polluting the database with this table.

It is true though that, because this object stores the DB connection, it is not thread-safe (no choices in this function could remedy that, I believe).  Multiple simultaneous instances are safe, and cannot see each other's temporary tables.

Reply:
I would hope Python raises an exception instead of crashing! But if you're confident the `CREATE TABLE` command is included, that's good enough for me.

Most Stack classes are not thread-safe, but AFAIK our parallelism is based on thread isolation, so that shouldn't be a problem.